### DISCUSSION TITLE
Combine documents

### DISCUSSION THREAD

**Discussion by markgould** [QuestPDF/QuestPDF#365] [Q&A] [answered]
What would be the best approach to combine multiple documents without resorting to using something like PdfSharp after the fact?

I'd be happy to try and figure out a way to do this but wanted to see if this is already possible and I'm just missing something.

**Comment by bennetbo** [comment]
No not possible currently and i believe not planned either.
QuestPDF uses SkiaSharp for PDF Generation, so QuestPDF basically draws on a canvas and Skia handles the translation to PDF commands.

To combine or modify PDFs, we would need to write a PDF parser/generator which is a VERY ambitious project itself.

**Reply by markgould** [reply]
Just to clarify, I am not looking to combine existing PDFs. To give a simple example, lets say I have an InvoiceDocument that takes an invoice model. Perhaps I want to generate 10 invoices into one PDF. Or perhaps I have documents of different types in QuestPDF and rather than create individual files I would like to render them together into one document.

**Reply by bennetbo** [reply]
Oh i see, i guess you could do something like this:
```csharp
public class MergedDocument : IDocument
{
    private readonly DocumentMetadata _metadata;
    private readonly IDocument[] _documents;

    public MergedDocument(DocumentMetadata metadata, params IDocument[] documents)
    {
        _metadata = metadata;
        _documents = documents;
    }

    public void Compose(IDocumentContainer container)
    {
        foreach (var doc in _documents)
            doc.Compose(container);
    }

    public DocumentMetadata GetMetadata() => _metadata;
}

//Usage
var doc = Document.Create(....);
var doc2 = Document.Create(...);

new MergedDocument(DocumentMetadata.Default, doc, doc2).GeneratePdf(....);
```

Although i think it would be nicer to build this into the library. Something like:
`Document.Merge(doc1, doc2).WithMetadata(...).GeneratePdf(...)`.

Edit: One problem with this approach is, that the page numbers will be messed up (wont be reset between 'seperate' documents, because it will now be treated as one).

**Reply by bennetbo** [reply]
@markgould Just opened a PR for this (#367). This PR also solves the page numbering problem i mentioned in my last comment.
Let me know what you think.

**Reply by markgould** [reply]
This looks excellent, thank you for adding that so quickly! The page numbering flexibility is great.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

The functionality you described is now available through the `DocumentOperation` API, introduced in version 2024.12.0. You can load and merge existing PDF files natively. Full documentation is available here:
https://www.questpdf.com/concepts/document-operations.html

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
